keps/sig-storage/2546-reject-invalid-pvc-datasource/README.md

```diff
@@ -0,0 +1,286 @@
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+[Volume Populators KEP]: ../1495-volume-populators/README.md
+
+While developing the [Volume Populators KEP] it became obvious that the way the admission
+controller handles the `DataSource` field of PVC specs is contrary to everyone's intentions.
+
+The existing behavior, which has been in place since 1.12, is to clear the field if the user
+supplies a value that is not one of the supported values. This results in Kubernetes behaving
+as if the user requested an empty volume, and indeed the user is given an empty volume in
+that case (absent any other errors).
+
+Developers never intended the `DataSource` field to work this way, and it's clearly not good
+practice to ignore user input. Most likely this was an accidental behavior resulting from the
+need to support disabling feature gates for specific kinds of data sources.
+
+Users cannot possibly depend on this behavior, because for every invalid request that we are
+currently incorrectly treating as valid, there is an actual valid request that results in the
+exact same result -- specifically a PVC with an empty data source.
+
+## Motivation
+
+### Goals
+
+- Fix the buggy admission controller behavior
+- Generate a clearly visible release note for users in the unlikely event that anyone is
+  affected.
+
+### Non-Goals
+
+- This change makes sense regardless of the future of the Volume Populators KEP. As such
+  is it not a goal of this KEP to advance the Volume Populators feature, although both
+  KEPs together to make good sense.
+
+## Proposal
+
+We propose to correct this bad behavior by making the admission controller reject invalid
+data sources instead of silently changing them. This would cause the admission controller
+to reject PVCs that specify core object other than PVCs as data sources when the
+AnyVolumeDataSource feature gate is enabled, and to reject PVCs that specify anything
+other than the two supported data sources (PVCs and VolumeSnapshots) when the feature gate
+is disabled.
+
+No existing workflows would be affected by this change, unless they are accidentally
+specifying a data source when they wanted an empty volume. Correcting those workflows would
+be trivial -- users would simply need to remove the data source if they wanted an empty
+volume. Valid data sources would be completely unaffected.
+
+### Risks and Mitigations
+
+The main risk is if a user had some preconfigured workflow that involved creation of
+PVCs with the `DataSource` field set to an invalid value. The workflow would be yielding
+empty volumes today, and if that was acceptable (most likely because the user forgot
+they ever put contents in the `DataSource` field) then it could go on unnoticed. Fixing
+this bug will cause that user's workflow to suddenly break.
+
+The workaround for the user would be trivial, they would just need to clear the
```

Comment:
Depending on where the request is coming from, it might be:
* trivial (manifest under the user's control)
* possible (programmatic client that needs to be updated)
* difficult (programmatic client not under the user's control that needs to be updated)
* impossible (already-persisted statefulset with a spec.volumeClaimTemplates with an invalid data source)

What is the recourse for users on the more difficult end of the scale? How much time do we give them to move?